### PR TITLE
codeintel: LSIF line reader

### PR DESCRIPTION
**Pull request by efritz**
Refactor the function that reads and unmarshalls lines of LSIF data from a source. This is done in a generic (focus on reading, not unmarshalling) way so we can replace/supplment JSON as an encoding in the future.

There are three significant changes: 

- Add a buffered channel between the bufio Scanner and the consumer so that IO can be done ahead of the consumer
- Introduce a pool of unmarshal workers that can work on a batch of lines concurrently
- Buffer the output channel so that unmarshalling can be done ahead of the correlator process

Running the precise-code-intel integration suite's upload phase on master:

```
3m37.597s
3m35.858s
3m42.591s
```

And after this change:

```
2m29.395s
2m32.812s
2m30.429s
```

### CODE REVIEW

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+					done = true
+					break
+				}
+
+				lines[i] = line
+				work <- i
+				i++
+			}
+
+			// Wait until the current batch has bee completely unmarshalled
```
Comment (ryanslade) [inline cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:94]: Typo, bee

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			}
+
+			// Read the result array in order. If the caller context has completed,
+			// we'll abandon any additional values we were going to send on this
+			// channel (as well as any additional errors from the scanner).
+			for j := 0; j < i; j++ {
+				select {
+				case pairCh <- pairs[j]:
+				case <-ctx.Done():
+					return
```
Comment (ryanslade) [inline cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:106]: After this return nothing will be reading from `lineCh` which will cause the scanner goroutine to leak if there are more than 512 lines to read.

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	scanner.Buffer(make([]byte, LineBufferSize), LineBufferSize)
+
+	// Read the document in a separate go-routine.
+	lineCh := make(chan []byte, ChannelBufferSize)
+	go func() {
+		defer close(lineCh)
+
+		for scanner.Scan() {
+			if line := scanner.Bytes(); len(line) != 0 {
+				lineCh <- line
```
Comment (keegancsmith) [inline cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:39]: @efritz You can't process lines from a Scanner concurrently. The slice returned is directly from the underlying buffer, so is mutated as more data is read. See godoc for Bytes:

> The underlying array may point to data that will be overwritten by a subsequent call to Scan.

Comment (efritz) [reply cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:39]: I'll ensure it's copied before being put on the channel.

Comment (keegancsmith) [reply cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:39]: Is that worth the cost in memory pressure for concurrency? I would suspect no to be honest. But probably better to copy and rely on empirical evidence after something goes wrong and your profile.

Comment (efritz) [reply cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:39]: The profile I used to make _this_ changed showed that the vast majority of the time was spent in json and doing _that_ concurrently is the big win (it drops 30% of conversion time off of each upload).

(Temporary) fix here: https://github.com/sourcegraph/sourcegraph/pull/11088

I'm re-profiling to ensure that it's not a huge regression (and we can always use an easy byte pool if it is).

**Pull request merged**